lib/modules/manager/gomod/package-tree.ts

```diff
@@ -0,0 +1,187 @@
+import { Graph, hasCycle } from 'graph-data-structure';
+import upath from 'upath';
+import { logger } from '../../../logger';
+import { readLocalFile } from '../../../util/fs';
+import { newlineRegex, regEx } from '../../../util/regex';
+import { scm } from '../../platform/scm';
+
+export interface GoModuleFile {
+  name: string;
+  isLeaf: boolean;
+}
+
+const replaceRegex = regEx(
+  /^replace\s+(?<module>[^\s]+)\s*=>\s*(?<replacement>[^\s]+)(?:\s+(?<version>[^\s]+))?(?:\s*\/\/.*)?$/,
+);
+
+/**
+ * Get all dependent go.mod files that need to be updated when packageFileName changes
+ */
+export async function getTransitiveDependentModules(
+  packageFileName: string,
+): Promise<GoModuleFile[]> {
+  const goModFiles = await getAllGoModFiles();
+
+  // If the target file doesn't exist in the repository, return empty
+  if (!goModFiles.includes(packageFileName)) {
+    return [];
+  }
+
+  const graph = new Graph();
+
+  for (const f of goModFiles) {
+    graph.addNode(f);
+  }
+
+  for (const f of goModFiles) {
+    const replaceDirectives = await parseLocalReplaceDirectives(f);
+
+    for (const directive of replaceDirectives) {
+      const referencedGoMod = directive.targetGoModPath;
+      // Only add edges for files that exist in our go.mod files list
+      if (referencedGoMod && goModFiles.includes(referencedGoMod)) {
+        // Create dependency edge: f depends on referencedGoMod
+        graph.addEdge(referencedGoMod, f);
+      }
+    }
+  }
+
+  if (hasCycle(graph)) {
+    logger.warn('Circular reference detected in Go modules replace directives');
+    return [];
+  }
+
+  const visitedModules = new Map<string, boolean>();
+  recursivelyGetDependentGoModFiles(packageFileName, graph, visitedModules);
+
+  // deduplicate
+  return Array.from(visitedModules).map(([name, isLeaf]) => ({ name, isLeaf }));
+}
+
+/**
+ * Traverse graph and find dependent go.mod files at any level of ancestry
+ */
+function recursivelyGetDependentGoModFiles(
```

Comment:
also shared, see below. i think it needs a separate refactor to move from nuget to utils.